File: Leetcode/03_Hash-Table/017_0202_happy-number.py
# https://leetcode.cn/problems/happy-number
# e

class Solution(object):
    def isHappy(self, n):
        seen = set()
        while n != 1:
            n = str(n)
            s = 0
            for d in n:
                s += int(d) ** 2
            if s in seen:
                return False
            else:
                seen.add(s)
                n = s
        return True

# 快慢指针
class Solution(object):
    def isHappy(self, n):
        f = s = n
        while f != 1:
            f = self.get_next(self.get_next(f))
            s = self.get_next(s)
            if s == f and s != 1:
                return False
        return True

    def get_next(self, num):
        s = 0
        num = str(num)
        for d in num:
            s += int(d) ** 2
        return s

File: Leetcode/03_Hash-Table/017_0202_happy-number.go
// https://leetcode.cn/problems/happy-number
// e

package main

func isHappy1(n int) bool {
	seen := map[int]struct{}{}
	for n != 1 {
		sum := 0
		for n > 0 {
			sum += (n % 10) * (n % 10)
			n = n / 10
		}
		if _, exists := seen[sum]; exists {
			return false
		}
		seen[sum] = struct{}{}
		n = sum
	}
	return true
}

func isHappy2(n int) bool {
	fast, slow := n, n
	for fast != 1 {
		fast = getNext(getNext(fast))
		slow = getNext(slow)
		if fast == slow && fast != 1 {
			return false
		}
	}
	return fast == 1
}

func getNext(n int) (sum int) {
	for tmp := n; tmp > 0; {
		sum += (tmp % 10) * (tmp % 10)
		tmp /= 10
	}
	return
}
